File: src/main/java/com/abysalto/middleware/config/UserSeeder.java
package com.abysalto.middleware.config;

import com.abysalto.middleware.model.Role;
import com.abysalto.middleware.model.UserAccount;
import com.abysalto.middleware.repository.UserRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.boot.ApplicationArguments;
import org.springframework.boot.ApplicationRunner;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Component;

/**
 * Creates the configured seed user on startup (when enabled and not already present) so the API
 * can be exercised without a manual registration step. The password is stored only as a BCrypt hash.
 */
@Component
public class UserSeeder implements ApplicationRunner {

    private static final Logger log = LoggerFactory.getLogger(UserSeeder.class);
    private static final Role DEFAULT_ROLE = Role.USER;

    private final SeedUserProperties props;
    private final UserRepository users;
    private final PasswordEncoder passwordEncoder;

    public UserSeeder(SeedUserProperties props, UserRepository users, PasswordEncoder passwordEncoder) {
        this.props = props;
        this.users = users;
        this.passwordEncoder = passwordEncoder;
    }

    @Override
    public void run(ApplicationArguments args) {
        if (!props.enabled()) {
            return;
        }
        if (users.existsByUsername(props.username())) {
            log.info("Seed user '{}' already exists; skipping.", props.username());
            return;
        }
        UserAccount user = new UserAccount(
                props.username(),
                passwordEncoder.encode(props.password()),
                DEFAULT_ROLE);
        users.save(user);
        log.info("Seeded user '{}' with role {}.", props.username(), DEFAULT_ROLE);
    }
}

File: dotnet/src/Middleware.Infrastructure/Security/UserSeeder.cs
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;
using Middleware.Core.Domain;
using Middleware.Core.Options;
using Middleware.Infrastructure.Persistence;

namespace Middleware.Infrastructure.Security;

/// <summary>
/// Creates the configured seed user on startup (when enabled and not already present) so the API can
/// be exercised without a manual registration step. The password is stored only as a BCrypt hash.
///
/// <para>Registered after <see cref="DatabaseInitializer"/> so the schema exists before the insert:
/// hosted services start sequentially in registration order.</para>
/// </summary>
public sealed class UserSeeder(
    IServiceScopeFactory scopeFactory,
    IOptions<SeedUserOptions> options,
    IPasswordHasher passwordHasher,
    ILogger<UserSeeder> logger) : IHostedService
{
    private const Role DefaultRole = Role.User;

    public async Task StartAsync(CancellationToken cancellationToken)
    {
        var seed = options.Value;
        if (!seed.Enabled)
        {
            return;
        }

        using var scope = scopeFactory.CreateScope();
        var users = scope.ServiceProvider.GetRequiredService<IUserRepository>();

        if (await users.ExistsByUsernameAsync(seed.Username, cancellationToken))
        {
            logger.LogInformation("Seed user '{Username}' already exists; skipping.", seed.Username);
            return;
        }

        var user = new UserAccount(seed.Username, passwordHasher.Hash(seed.Password), DefaultRole);
        await users.AddAsync(user, cancellationToken);
        logger.LogInformation("Seeded user '{Username}' with role {Role}.", seed.Username, DefaultRole);
    }

    public Task StopAsync(CancellationToken cancellationToken) => Task.CompletedTask;
}
